### PR TITLE
Triggerevent

### LINKED ISSUE #1314: codeworld-api: NotReady instance for ReflexCodeWorld
It would be neat to provide a `NotReady` instance among the other reflex constraints, so that `networkHold` is available in reflex codeworld-api programs.

### LINKED ISSUE #1328: Add TriggerEvent to ReflexCodeWorld
I am working on a framework for simple two player games, as part of a general-purpose assignment skeleton. I want to provide hooks for a CodeWorld interface to the games. I had about 80% of a solution coded up, where values are passed to and from the game engine by `MVar`s, but I hit a wall when the lack of `TriggerEvent` prevented me from writing `watchMVar :: (TriggerEvent t m, PerformEvent t m) => MVar a -> m (Event t a)`.

(This is the same as my remark in #1053, but after sleeping on it, I decided that it deserved an open issue, and I don't have privs to reopen issues in this repo.)

### PR DESCRIPTION
**Pull request by endgame**
Add `TriggerEvent` instance to `ReflexCodeWorld`

Where do you get your ghcjs from? I have built this branch using the one from reflex-platform, but it seemed to be missing an aeson marshaling function. I replaced it with a call to `undefined` just to make things build, but didn't commit that to this branch.

I have tested the code under native ghc, and it appears to work.

I also added a `MonadIO` instance, since it didn't seem to do much harm. Given that we have `MonadIO (Performable m)`, reflex users can do arbitrary IO anyway.

also tidied things up a bit on the way through

Closes #1328 
Closes #1314

### CODE REVIEW

**Comment by endgame** [comment]
@googlebot I signed it!

File: codeworld-api/src/CodeWorld/Driver.hs [hunk 1]
```diff
@@ -1905,7 +1955,7 @@ createPhysicalReactiveInput window canvas fire = do
 
     pointerPosition <- R.holdDyn (0, 0) pointerMovement
     pointerDown <- R.holdDyn False $
-        R.mergeWith (&&) [True <$ pointerPress, False <$ pointerRelease]
+        R.leftmost [True <$ pointerPress, False <$ pointerRelease]
```
[the remark this answers is not in the document]
Comment (endgame) [reply codeworld-api/src/CodeWorld/Driver.hs:1958]: You're right, it changes things. I replaced some of the other uses of `mergeWith` that were equivalent to `leftmost`, and mistakenly thought that this was a change with no effect.

[the remark this answers is not in the document]
Comment (endgame) [reply codeworld-api/src/CodeWorld/Driver.hs:1958]: As you say, I don't think it makes a difference given that you can't get the pointer down and up in the same frame.

**Comment by endgame** [comment]
Thanks for reviewing and merging this so quickly. Just for my planning's sake, do you have a rough sense of when you plan to push the next release to hackage? I just want to know whether to look at using `cabal.project` files to point at my fork or whatever if it's not coming soon.

**Comment by cdsmith** [comment]
I will try to verify and push this tonight.

**Comment by cdsmith** [comment]
I'm actually not able to build this on the production server.  It builds fine with GHCJS, but with native GHC, I get:

```
src/CodeWorld/Driver.hs:2109:13: error:
    • No instance for (R.NotReady
                         (R.SpiderTimeline R.Global)
                         (R.PerformEventT
                            (R.SpiderTimeline R.Global) (R.SpiderHost R.Global)))
        arising from a use of ‘program’
    • In the second argument of ‘($)’, namely ‘program input’
      In a stmt of a 'do' block:
        (_inspectPicture, displayPicture) <- flip
                                               R.runTriggerEventT triggeredEvents
                                               . flip R.runPostBuildT postBuild
                                               $ program input
      In the second argument of ‘($)’, namely
        ‘do (_inspectPicture, displayPicture) <- flip
                                                   R.runTriggerEventT triggeredEvents
                                                   . flip R.runPostBuildT postBuild
                                                   $ program input
            R.performEvent_
              $ liftIO
                  <$>
                    R.leftmost
                      [frame <$> R.updated displayPicture,
                       frame <$> R.tagPromptlyDyn displayPicture postBuild]
            return ()’
     |
2109 |             program input
     |             ^^^^^^^^^^^^^
```

**Comment by endgame** [comment]
Well that's annoying. What version of reflex is being selected on your prod machines? The Reflex changelog says that the `instance NotReady (SpiderTimeline x) (SpiderHost x)` doesn't come in until 0.6.3.

**Comment by cdsmith** [comment]
Okay, great!  I've updated the version constraints and will try to rebuild.  It must be that the native GHC package database had a different installed version, and the right version bounds should fix it.

**Comment by cdsmith** [comment]
I have released codeworld-api 0.5 with these changes.

**Comment by endgame** [comment]
Great, thanks very much.

**Pull request merged**